### ISSUE TITLE
Several Question about IID Segmentation

### ISSUE THREAD

**Issue by kongyanye** [xu-ji/IIC#42]
Hello Xu, thanks for the great work! I'm adopting your approach in my unsupervised segmentation task and encountered several problems about the scripts.

1) About the 'local spatial invariance' technique in Section 3.3, where's the code implementing it? In dataloader of coco-stuff-3 and potsdam, the paired patches only contains photometric transformation (one raw img, and the other is color jitter version) instead of pairs with shifts. See [here](https://github.com/xu-ji/IIC/blob/ce64bd2dc9dbbeaff270550e393bcaa51c5a8166/code/datasets/segmentation/cocostuff.py#L174).

2) I'm implementing unsupervised image segmentation for RGB images, which should be the same as ['Fully unsupervised segmentation for coco-stuff-3'](https://github.com/xu-ji/IIC/blob/ce64bd2dc9dbbeaff270550e393bcaa51c5a8166/examples/commands.txt#L74). After checking the code, I found the ground truth [label/mask](https://github.com/xu-ji/IIC/blob/ce64bd2dc9dbbeaff270550e393bcaa51c5a8166/code/datasets/segmentation/cocostuff.py#L110) is loaded in dataloader  and [used for computing loss in the training phase](https://github.com/xu-ji/IIC/blob/ce64bd2dc9dbbeaff270550e393bcaa51c5a8166/code/scripts/segmentation/segmentation_twohead.py#L321). And in the dataloader for Potsdam, the corresponsive label/mask is [set to all ones](https://github.com/xu-ji/IIC/blob/ce64bd2dc9dbbeaff270550e393bcaa51c5a8166/code/datasets/segmentation/potsdam.py#L207), which I think makes sense. So which should be the right choice for mask_img1 in unsupervsed segmentation task?

3) I noticed that the function pad_and_or_crop is used both for training and testing part in dataloader. Does cropping images into size (h,w) performs better than directly resizing them into (h,w)? And if you crop them when training, to obtain the whole segmentation map for testing images, we need to generate masks using sliding window with size (h,w) instead of just [center crop the img](https://github.com/xu-ji/IIC/blob/ce64bd2dc9dbbeaff270550e393bcaa51c5a8166/code/datasets/segmentation/potsdam.py#L318) to (h,w), right?

Please correct me if there's any misunderstanding. Thanks for your time!

**Comment by xu-ji** [comment]
Hi,
1. Local spatial invariance is done in the loss function, see [here](https://github.com/xu-ji/IIC/blob/master/code/utils/segmentation/IID_losses.py#L125). In the loss function we take each pixel prediction and match it up to every other pixel prediction in its local neighbourhood box, and this is implemented as a convolution. This is discussed in the "Implementation" paragraph in sec. 3.3. of the paper.
2. Normally, if you want every pixel in your dataset to be used, you wouldn't use any masks (i.e. same as Potsdam). The masking is done for COCO-stuff purely to remove the non-stuff pixel predictions from training and evaluation.
3.  I think only in the case of COCO-stuff is there actually a crop; Potsdam's input size is already 200x200 so pad_and_or_crop(img, 200) shouldn't do anything to it. In the case of the former, the cropping is used to get all images into the same dimensions without changing aspect ratio. For this reason I didn't test resizing. The network operates on a fixed input size. So you're right, we could have done multiple passes on each image to get the full segmentation map but we chose just to do the centre (for us and all baselines). Since we already shrink each image to a third of its size, 200x200 captures most of the original image.

**Comment by kongyanye** [comment]
Thanks for the clarification. It helps a lot.
